### PR TITLE
added new macros, routine and diagram computation

### PR DESCRIPTION
**Pull request by roberto-dionisio**

### CODE REVIEW

File: src/measures/Makefile.am [hunk 1]
```diff
@@ -21,7 +21,8 @@ __top_builddir__lib_libnissa_a_SOURCES+= \
 	%D%/gauge/all_rectangles.cpp \
 	%D%/gauge/pline.cpp \
 	%D%/gauge/topological_charge.cpp \
-	%D%/gauge/watusso.cpp
+	%D%/gauge/watusso.cpp
```
Comment (sunpho84) [inline src/measures/Makefile.am:24]: trailing space to be removed

File: src/measures/Makefile.am [hunk 2]
```diff
@@ -21,7 +21,8 @@ __top_builddir__lib_libnissa_a_SOURCES+= \
 	%D%/gauge/all_rectangles.cpp \
 	%D%/gauge/pline.cpp \
 	%D%/gauge/topological_charge.cpp \
-	%D%/gauge/watusso.cpp
+	%D%/gauge/watusso.cpp 
+
```
Comment (sunpho84) [inline src/measures/Makefile.am:25]: empty line to be removed

File: src/nissa.hpp [hunk 3]
```diff
@@ -137,6 +137,7 @@
 #include "measures/fermions/tm_tuning.hpp"
 #include "measures/fermions/zumba.hpp"
 
+
```
Comment (sunpho84) [inline src/nissa.hpp:140]: empty line to be removed

File: src/measures/fermions/ellesettete.cpp [hunk 4]
```diff
@@ -87,131 +84,55 @@ namespace nissa
...
+				MINV(SIMPLE_PROP,iflav,source);			
+				
+				//compute sequential propagator G(m|n) ~ [D^-1(m|y) source(y)] D^-1(y|n)
+				MINV(SEQ_PROP,iflav,SIMPLE_PROP); //
+				
+				//then glb reduction to compute the trace for the connected 3pts diagram 
+				SUMM_THE_TRACE_PRINT_AT_LAST_HIT(Tr_three_pts,SIMPLE_PROP,SEQ_PROP);
+				
+				//////// disconnected //////// 
+				PUT_G5G5_STAG_PHASES_WITH_NO_SHIFT(g5_source,iflav,source)
```
[the remark this answers is not in the document]
Comment (roberto-dionisio) [reply src/measures/fermions/ellesettete.cpp:112]: I wanted to preserve the good old style of the code

File: src/measures/fermions/stag.cpp [hunk 5]
```diff
@@ -162,6 +162,21 @@ namespace nissa
 	  set_borders_invalid(source[eo]);
 	}
     }
+
+
+	void put_g5g5_stag_phases_with_no_shift(eo_ptr<color> out,int iflav,eo_ptr<color> in)
```
[the remark this answers is not in the document]
Comment (roberto-dionisio) [reply src/measures/fermions/stag.cpp:167]: do u mean ispin and itaste?

[the remark this answers is not in the document]
Comment (sunpho84) [reply src/measures/fermions/stag.cpp:167]: that's what I meant, however, in that case the routine would not be called `put_g5g5_stag_phase_with_no-shift` but something like `put_stag_op`, moreover one would not add any g5g5 by default and would let the user pass exactly the required operator. The `form_stag_meson_pattern_with_g5g5` must be replaced with `form_stag_meson_pattern`

[the remark this answers is not in the document]
Comment (roberto-dionisio) [reply src/measures/fermions/stag.cpp:167]: `apply_stag_op` is already implemented in stag.hpp. That, given a `ispin` and `itaste`, forms the mask, the shift and then applies both phases and shift to the operator.

[the remark this answers is not in the document]
Comment (roberto-dionisio) [reply src/measures/fermions/stag.cpp:167]: and then, if we call `apply_stag_op` with 15 for both spin and taste, that automatically gives the desired stagg phases and a null shift. No?

File: src/measures/fermions/ellesettete.cpp [hunk 6]
```diff
@@ -108,11 +108,11 @@ namespace nissa
 				SUMM_THE_TRACE_PRINT_AT_LAST_HIT(Tr_three_pts,SIMPLE_PROP,SEQ_PROP);
 				
 				//////// disconnected //////// 
-				apply_stag_op(g5_source,conf,theory_pars->backfield[iflav],source);
+				apply_stag_op(g5_source,conf,theory_pars->backfield[iflav],15,15,source);
```
Comment (sunpho84) [inline src/measures/fermions/ellesettete.cpp:111]: 15 needs to be replaced with: GAMMA_INT::GAMMA_5

File: src/measures/fermions/ellesettete.cpp [hunk 7]
```diff
@@ -108,11 +108,11 @@ namespace nissa
...
 				
 				//MINV(G5_PROP,iflav,g5_source); 
 				SUMM_THE_TIME_TRACE_PRINT_AT_LAST_HIT(Tr_first_bubble, SEQ_PROP, g5_source);
```
Comment (sunpho84) [inline src/measures/fermions/ellesettete.cpp:114]: Tr_first_bubble is a complex vector whereas the macro calls a routine which expects a double vector, you must change one of the two

File: src/measures/fermions/ellesettete.cpp [hunk 8]
```diff
@@ -87,131 +84,54 @@ namespace nissa
...
-    DELETE_FIELD_T(d2M_M);
-    DELETE_FIELD_T(M_M);
-    DELETE_FIELD_T(dM_M_M);
-    DELETE_FIELD_T(d2M_M_M);
-    DELETE_FIELD_T(dM_M_dM_M);
-    DELETE_FIELD_T(M_dM_M_dM_M);
-    DELETE_FIELD_T(TMP);
+    DELETE_FIELD_T(SIMPLE_PROP);
+    DELETE_FIELD_T(SEQ_PROP);
+    DELETE_FIELD_T(G5_PROP);
```
Comment (sunpho84) [inline src/measures/fermions/ellesettete.cpp:126]: G5_PROP is not allocated

File: src/measures/fermions/stag.cpp [hunk 9]
```diff
@@ -212,24 +213,43 @@ namespace nissa
...
+	
+	void summ_the_time_trace(double* out,complex* point_result,eo_ptr<color>  A,eo_ptr<color>  B)
+    {
+      
+      //compute results for single points
+      local_trace(point_result, A, B);
+      
+      // reduction over 3-spatial-volume
+      complex temp[glbSize[0]];
+      glb_reduce(temp,point_result,locVol,glbSize[0],locSize[0]);	//(ask sunpho)check if the parameters passed for the reduction over space are correct
```
Comment (sunpho84) [inline src/measures/fermions/stag.cpp:248]: you have to add ,glbCoordOfLoclx[0][0] as a last argument as in https://github.com/sunpho84/nissa/blob/827ca27f9c794b9a5c3eb0eab50aaf2a2c09eca2/src/measures/fermions/mesons.cpp#L116

File: src/measures/fermions/stag.hpp [hunk 10]
```diff
@@ -59,19 +59,30 @@ namespace nissa
...
+      master_fprintf(file," # %s flav %d\n",#A,iflav);    \
+      for(int t=0;t<glbSize[0];t++)                     \
+      master_fprintf(file,"%+16.16lg %+16.16lg\t",A[t][0]/meas_pars.nhits,A[t][1]/meas_pars.nhits);         \
+      master_fprintf(file,"\n")                     
 #define SUMM_THE_TRACE_PRINT_AT_LAST_HIT(A,B,C)				\
       summ_the_trace((double*)A,point_result,B,C);			\
       if(ihit==meas_pars.nhits-1) PRINT(A)
-      
+#define SUMM_THE_TIME_TRACE_PRINT_AT_LAST_HIT(A,B,C) \
+      summ_the_time_trace((double*)A,point_result,B,C);	\
```
Comment (sunpho84) [inline src/measures/fermions/stag.hpp:71]: the cast to double* can be removed as this is a double vector

File: src/measures/fermions/stag.hpp [hunk 11]
```diff
@@ -59,19 +59,30 @@ namespace nissa
     delete[] NAME2(loc,o)
 #define PRINT(A)							\
       master_fprintf(file,"%+16.16lg %+16.16lg\t",A[0]/meas_pars.nhits,A[1]/meas_pars.nhits)
+#define PRINT_VEC(A)							\
+      master_fprintf(file," # %s flav %d\n",#A,iflav);    \
+      for(int t=0;t<glbSize[0];t++)                     \
+      master_fprintf(file,"%+16.16lg %+16.16lg\t",A[t][0]/meas_pars.nhits,A[t][1]/meas_pars.nhits);         \
```
Comment (sunpho84) [inline src/measures/fermions/stag.hpp:65]: here, a complex is expected... maybe you could write two routines `print_vec_impl` (implementation), one taking a double vector and one taking a complex

Comment (sunpho84) [reply src/measures/fermions/stag.hpp:65]: of course one could have a print_vec routine by itself....

Comment (roberto-dionisio) [reply src/measures/fermions/stag.hpp:65]: but we only need the one for the double case... So  I'm gonna create two routines `print_vec_complex` and `print_vec_double` anyway, and at the place of `PRINT_VEC` in `SUMM_THE_TIME_TRACE_PRINT_AT_LAST_HIT` we're gonna call  `print_vec_double`, right?

Comment (sunpho84) [reply src/measures/fermions/stag.hpp:65]: you mean that PRINT_VEC is used only in this situation, I see. So you can directly assume it's a real vector without the need to tell apart the case of complex. If needed in the future we can write the case of a complex

Comment (roberto-dionisio) [reply src/measures/fermions/stag.hpp:65]: should I pass to the `print_vec_double` the generic size for the array one passes or just one argument for our specific case?

Comment (sunpho84) [reply src/measures/fermions/stag.hpp:65]: if PRINT_VEC is not used anywhere else, you can adapt the macro as you want

**Pull request merged**